### PR TITLE
ramips: add support for HiWiFi HC5611

### PR DESCRIPTION
**Pull request by nanpuyue**
## Hardware:
CPU: MediaTek MT7628AN @ 575 MHz
Flash: 16 MB
RAM: 128 MB
Ethernet: 10/100Mbps x 1
Wlan: 300 Mbps
USB: USB 2.0 x 1
LED: red/green x 1
Button: reset x 1

## Installation:
1. Apply for developer mode on https://app.hiwifi.com
2. Install the developer mode plugin
3. Upload the openwrt firmware to the router via SCP
4. Login the router via SSH
5. Run `mtd -r write path_to_firmware.bin firmware`

### CODE REVIEW

**Comment by pepe2k** [comment]
Hello @nanpuyue,

Please, read our [submitting patches rules](https://lede-project.org/submitting-patches), and make sure your patch or pull request follows them. Please, use full name in SoB line and include hardware/installation information in description of your commit.

Cheers,
Piotr

**Comment by nanpuyue** [comment]
Done

**Comment by pepe2k** [comment]
> Done

Nope.

**Comment by mkresin** [comment]
Beside that fact that neither the hardware description nor the install instructions are in the **commit message**, this

> mtd -r write openwrt-ramips-mt76x8-hc5611-squashfs-sysupgrade.bin firmware

is not what I consider as install instruction from stock firmware. Please have a look at the [recent additions](https://git.openwrt.org/?p=openwrt%2Fopenwrt.git&a=search&h=HEAD&st=commit&s=add+support+for) for some better examples.

**Comment by nanpuyue** [comment]
Done again 😂

**Comment by mkresin** [comment]
Before I do the real review, the LED is a dual colour LED, attached to two GPIOs?

**Comment by nanpuyue** [comment]
@mkresin 
Yes，gpio6 & gpio11.

File: target/linux/ramips/base-files/etc/board.d/01_leds [hunk 1]
```diff
@@ -205,6 +205,9 @@ hc5661a)
 	ucidef_set_led_netdev "internet" "internet" "$boardname:blue:internet" "eth0.2"
 	set_wifi_led "$boardname:blue:wlan2g"
 	;;
+hc5611)
+	ucidef_set_led_default "system" "system" "$boardname:green:system" "1"
```
Comment (mkresin) [inline target/linux/ramips/base-files/etc/board.d/01_leds]: Drop this LED here. the led is/will be handled by diag.sh and switched to ON after the boot is done.

File: target/linux/ramips/base-files/etc/diag.sh [hunk 2]
```diff
@@ -162,6 +162,7 @@ get_status_led() {
 		;;
 	hc5*61|\
 	hc5661a|\
+	hc5611|\
```
Comment (mkresin) [inline target/linux/ramips/base-files/etc/diag.sh]: This one is wrong. You don't have a `blue:system` LED. Please use the `green:power` LED instead.

Furthermore, please use `hiwifi,hc5611` as boardname everywhere.

File: target/linux/ramips/base-files/lib/ramips.sh [hunk 3]
```diff
@@ -223,6 +223,9 @@ ramips_board_detect() {
 	*"HC5661A")
 		name="hc5661a"
 		;;
+	*"HC5611")
```
Comment (mkresin) [inline target/linux/ramips/base-files/lib/ramips.sh]: drop this to use the generic board detection. The userspace boardname will be `hiwifi,hc5611`.

File: target/linux/ramips/dts/HC5611.dts [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+			gpios = <&gpio0 11 GPIO_ACTIVE_HIGH>;
+		};
+	};
+
+	gpio_export {
+		compatible = "gpio-export";
+		#size-cells = <0>;
+
+		power-led {
+			gpio-export,name = "power-led";
```
Comment (mkresin) [inline target/linux/ramips/dts/HC5611.dts]: Please add it as `hc5611:red:power` **LED** instead.

File: target/linux/ramips/dts/HC5611.dts [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+	memory@0 {
+		device_type = "memory";
+		reg = <0x0 0x8000000>;
+	};
+
+	gpio-leds {
+		compatible = "gpio-leds";
+
+		system {
+			label = "hc5611:green:system";
```
Comment (mkresin) [inline target/linux/ramips/dts/HC5611.dts]: Please name the LED `hc5611:green:power`,

File: target/linux/ramips/dts/HC5611.dts [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+			gpios = <&gpio1 6 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+	};
+};
+
+&pinctrl {
+	state_default: pinctrl0 {
+		gpio {
+			ralink,group = "i2c", "refclk", "wled_an";
```
Comment (mkresin) [inline target/linux/ramips/dts/HC5611.dts]: You are not using any of the pins covered by these groups as GPIO. It has to be `"spi_cs1", "wdt"`.

That the red:power LED uses the pin of the `spi_cs1` group looks rather strange. Are you sure it is correct?

File: target/linux/ramips/dts/HC5611.dts [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+&spi0 {
+	status = "okay";
+
+	m25p80@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		linux,modalias = "m25p80", "w25q128";
```
Comment (mkresin) [inline target/linux/ramips/dts/HC5611.dts]: drop this line. It isn required since any longer.

File: target/linux/ramips/image/mt76x8.mk [hunk 8]
```diff
@@ -41,6 +41,13 @@ define Device/hc5661a
 endef
 TARGET_DEVICES += hc5661a
 
+define Device/hc5611
```
Comment (mkresin) [inline target/linux/ramips/image/mt76x8.mk]: Please change it to `define Device/hiwifi_hc5611`

File: target/linux/ramips/image/mt76x8.mk [hunk 9]
```diff
@@ -41,6 +41,13 @@ define Device/hc5661a
...
+  DTS := HC5611
+  IMAGE_SIZE := $(ralink_default_fw_size_16M)
+  DEVICE_TITLE := HiWiFi HC5611
+endef
+TARGET_DEVICES += hc5611
```
Comment (mkresin) [inline target/linux/ramips/image/mt76x8.mk]: Please change it to `TARGET_DEVICES += hiwifi_hc5611`

File: target/linux/ramips/dts/HC5611.dts [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+	};
+};
+
+&ethernet {
+	mtd-mac-address = <&factory 0x4>;
+	mediatek,portmap = "wllll";
+};
+
+&wmac {
+	status = "okay";
```
Comment (mkresin) [inline target/linux/ramips/dts/HC5611.dts]: Have you tested the wireless? Quite surprising that no eeprom from the factory partition is required.

File: target/linux/ramips/dts/HC5611.dts [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+			label = "backup";
+			reg = <0xff0000 0x10000>;
+			read-only;
+		};
+	};
+};
+
+&ethernet {
+	mtd-mac-address = <&factory 0x4>;
+	mediatek,portmap = "wllll";
```
Comment (mkresin) [inline target/linux/ramips/dts/HC5611.dts]: drop this. If you only have one port, you don't need a portmap.

File: target/linux/ramips/dts/HC5611.dts [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+		partition@ff0000 {
+			label = "backup";
+			reg = <0xff0000 0x10000>;
+			read-only;
+		};
+	};
+};
+
+&ethernet {
+	mtd-mac-address = <&factory 0x4>;
```
Comment (mkresin) [inline target/linux/ramips/dts/HC5611.dts]: Something is wrong here. In 02_network you set the mac address from the `bdinfo` partition. But here you set a mac address from the factory partition.

Please compare the ethernet as well as wireless mac with the ones used by the stock firmware. Are they the same?

File: target/linux/ramips/image/mt76x8.mk [hunk 13]
```diff
@@ -41,6 +41,13 @@ define Device/hc5661a
 endef
 TARGET_DEVICES += hc5661a
 
+define Device/hc5611
+  DTS := HC5611
+  IMAGE_SIZE := $(ralink_default_fw_size_16M)
```
Comment (mkresin) [inline target/linux/ramips/image/mt76x8.mk]: ralink_default_fw_size_16M is 16121856 bytes (0xf60000) but your factory partition has a size of 0xf70000. It has to be:

```
IMAGE_SIZE := 15808k
```

**Review by mkresin — CHANGES_REQUESTED** [review]
Please test you image after you have done all the requested changes.

Please provide a full bootlog from an image with the requested changes applied.

File: target/linux/ramips/dts/HC5611.dts (hunk 4 above)
Comment (nanpuyue) [reply target/linux/ramips/dts/HC5611.dts]: They affect each other, how do I deal with this?

File: target/linux/ramips/image/mt76x8.mk (hunk 8 above)
Comment (nanpuyue) [reply target/linux/ramips/image/mt76x8.mk]: There is already hc5661 hc5661 hc5761 hc5861 hc5962 an so on, do I need to be consistent with them?
```bash
openwrt/target/linux/ramips$ grep -rin Device/hc5 image/*.mk
image/mt7620.mk:218:define Device/hc5661
image/mt7620.mk:226:define Device/hc5761
image/mt7620.mk:234:define Device/hc5861
image/mt7621.mk:86:define Device/hc5962
image/mt76x8.mk:37:define Device/hc5661a
image/mt76x8.mk:44:define Device/hc5611
```

File: target/linux/ramips/dts/HC5611.dts (hunk 6 above)
Comment (nanpuyue) [reply target/linux/ramips/dts/HC5611.dts]: HC5611.dts just be modified from HC5661A.dts, I can't confirm this.

File: target/linux/ramips/base-files/etc/diag.sh (hunk 2 above)
Comment (nanpuyue) [reply target/linux/ramips/base-files/etc/diag.sh]: It's my mistake.

File: target/linux/ramips/dts/HC5611.dts (hunk 4 above)
Comment (mkresin) [reply target/linux/ramips/dts/HC5611.dts]: Strange hardware. I'm not sure if I read this table correctly. Furthermore, I've no idea if 0 is LOW or HIGH.

Can you confirm my understanding of the table:

GPIO11 switches the power LED either on or off.
GPIO06 toggles the colour between red and green.

If my understanding is wrong, please provide the following information:

GPIO06 LOW + GPIO11 LOW =
GPIO06 LOW + GPIO11 HIGH =
GPIO06 HIGH + GPIO11 LOW =
GPIO06 HIGH + GPIO11 HIGH =

File: target/linux/ramips/dts/HC5611.dts (hunk 6 above)
Comment (mkresin) [reply target/linux/ramips/dts/HC5611.dts]: If the buttons and LEDs are still working, my proposed pinmux is correct. I had a look at the datasheet to check which GPIO belongs to which pinmux group.

File: target/linux/ramips/image/mt76x8.mk (hunk 8 above)
Comment (mkresin) [reply target/linux/ramips/image/mt76x8.mk]: No, you don't need to be consistent with them.

All newly added board will use the generic board detection. Existing boards are using a target specific board detection and will be migrated to the generic one within the next month.

File: target/linux/ramips/dts/HC5611.dts (hunk 4 above)
Comment (nanpuyue) [reply target/linux/ramips/dts/HC5611.dts]: GPIO06 LOW + GPIO11 LOW = RED
GPIO06 LOW + GPIO11 HIGH = NONE
GPIO06 HIGH + GPIO11 LOW = RED
GPIO06 HIGH + GPIO11 HIGH = GREEN

Comment (pepe2k) [reply target/linux/ramips/dts/HC5611.dts]: @nanpuyue are you able to provide PCB pictures, especially the part with that weird LEDs? And, are you sure that both GPIOs are properly exported as outputs?

Comment (nanpuyue) [reply target/linux/ramips/dts/HC5611.dts]: @pepe2k 
It's too difficult to open the case of the router, there is some pictures of the touter:
http://www.hiwifi.com/jgo-view
And I'm sure that both GPIOs are properly exported as outputs.

File: target/linux/ramips/dts/HC5611.dts (hunk 7 above)
Comment (nanpuyue) [reply target/linux/ramips/dts/HC5611.dts]: It doesn't work after drop.

File: target/linux/ramips/dts/HC5611.dts (hunk 4 above)
Comment (mkresin) [reply target/linux/ramips/dts/HC5611.dts]: > It's too difficult to open the case of the router, there are some pictures of the touter: http://www.hiwifi.com/jgo-view

Would you please provide us the pictures from the link. The website is down for me for the last two days. And according to https://www.uptrends.com/tools/uptime it is not only me.

I'm still curious how the hardware design looks like. Never seen such a strange config.

File: target/linux/ramips/dts/HC5611.dts (hunk 7 above)
Comment (mkresin) [reply target/linux/ramips/dts/HC5611.dts]: Than you have done something else wrong.  The devicetree property isn't used in kernel 4.9:

```
/dev/build_dir/target-mipsel_24kc_musl/linux-ramips_mt76x8/linux-4.9.77 $ grep -nir "linux,modalias"
Documentation/devicetree/bindings/spi/spi-fsl-dspi.txt:52:		linux,modalias = "m25p80"
/dev/build_dir/target-mipsel_24kc_musl/linux-ramips_mt76x8/linux-4.9.77 $
```

**Comment by mkresin** [comment]
Please reopen the PR if you have the time to fix the outlined issues.

File: target/linux/ramips/dts/HC5611.dts (hunk 4 above)
Comment (nanpuyue) [reply target/linux/ramips/dts/HC5611.dts]: https://www.right.com.cn/forum/thread-483273-1-1.html

Comment (nanpuyue) [reply target/linux/ramips/dts/HC5611.dts]: 图片
图片

**Comment by thank243** [comment]
@nanpuyue So, hc5611 have solved the outlined issues and reopen the PR?

**Pull request closed without merging**